### PR TITLE
[REFACTOR][TIR] Introduce PrimFuncPass.

### PR DESCRIPTION
**Pull request by tqchen**
- Introduce PrimFuncPass
- Convert one pass to the unified Pass API.

As part of the unified IR refactor. We will convert all TIR passes into tir.transform and tir.analysis namespace. The passes under tir.transform shares the same Pass infra as relay. Which takes a IRModule(with PrimFuncs) and return a updated IRModule

In order to keep things backward compatible, we will keep the old API around until most of the passes has been converted. Then we will do a bulk remove of the legacy API

### CODE REVIEW

**Comment by tqchen** [comment]
cc @yzhliu @kazum @vinx13 @merrymercy @Hzfengsy @ZihengJiang

File: python/tvm/tir/transform/function_pass.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+import tvm._ffi
+from tvm.ir.transform import Pass, PassInfo
+
+from . import _ffi_api
+
+
+@tvm._ffi.register_object("tir.PrimFuncPass")
+class PrimFuncPass(Pass):
+    """A pass that works on each tvm.relay.Function in a module. A function
```
Comment (zhiics) [inline python/tvm/tir/transform/function_pass.py]: tvm.tir.PrimFunc

File: python/tvm/tir/transform/function_pass.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+def prim_func_pass(pass_func=None, opt_level=None, name=None, required=None):
+    """Decorate a function pass.
+
+    This function returns a callback when pass_func
+    is provided. Otherwise, it returns the created function pass using the
+    given optimization function.
+
+    Parameters
+    ----------
+    pass_func : Optional[Callable[(Function, Module, PassContext) -> Function]]
```
Comment (zhiics) [inline python/tvm/tir/transform/function_pass.py]: Function->PrimFunc

Module->IRModule

File: python/tvm/tir/transform/function_pass.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    opt_level : int
+        The optimization level of this module pass.
+
+    name : Optional[str]
+        The name of the function pass. The name could be empty. In this case, the
+        name of the optimization function will be used as the pass name.
+
+    required : Optional[List[str]]
+        The list of passes that the module pass is dependent on.
```
Comment (zhiics) [inline python/tvm/tir/transform/function_pass.py]: function pass

File: src/tir/ir/transform.cc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("pass_info", &pass_info);
+  }
+
+  /*!
+   * \brief Run a function pass on given pass context.
+   *
+   * \param mod The module that an optimization pass is applied on.
+   * \param mod The context that an optimization pass executes on.
```
Comment (siju-samuel) [inline src/tir/ir/transform.cc]: mod -> pass_ctx

File: tests/python/unittest/test_tir_transform_combine_context_call.py [hunk 5]
```diff
@@ -37,9 +37,15 @@ def device_context(dev_id):
                     ("int32", "fadd", device_context(0), A))
     body = ib.get()
     f = tvm.tir.ir_pass.MakeAPI(body, "func", [dev_type, n], 2, True)
-    f = tvm.tir.ir_pass.CombineContextCall(f)
-    assert f.body.value.dtype == "handle"
-    assert f.body.body.value.dtype == "handle"
+
+    # temp adapter t convert loweredFunc to IRModule
```
Comment (siju-samuel) [inline tests/python/unittest/test_tir_transform_combine_context_call.py]: t -> to?

**Comment by tqchen** [comment]
Thanks @zhiics @siju-samue哈

**Pull request merged**